### DISCUSSION TITLE
Colorspace conversion using mogrify not working anymore?

### DISCUSSION THREAD

**Discussion by Joshua-Schroijen** [ImageMagick/ImageMagick#4959] [Help] [answered]
Hello ImageMagick community,  

Lately I noticed that the following command:  
`mogrify -colorspace YCbCr image.jpeg`  
does not perform a successful colorspace conversion from sRGB to YCbCr anymore. The colorspace remains unchanged after running  
`identify -verbose image.jpeg | grep Colorspace`

If I run  
`mogrify -debug -colorspace YCbCr image.jpeg`  
I get the following output:
`mogrify-im6.q16: unrecognized event type -colorspace' @ error/mogrify.c/MogrifyImageCommand/4528.`  

I know this issue exists since at least version 6.9.10-23 running on Ubuntu 20.04 LTS.

Am I invocating this command incorrectly these days or is this a bug?  

Thanks very much in advance,
Joshua

**Comment by snibgo** [comment]
JPEG files usually have pixels encoded as YCbCr, not sRGB.

When IM reads a JPEG file, it silently converts the pixels from YCbCr to sRGB.

When IM writes a JPEG, if the colorspace isn't already YCbCr, it silently converts the pixels to YCbCr.

**Reply by Joshua-Schroijen** [reply]
Thanks so much for your clarification!
But what I fail to understand is then what the Colorspace field means in the output of identify on a JPEG image?
If JPEG images processed by ImageMagick are always encoded in YCbCr unless explicitly requested otherwise, why is sRGB reported? I fear that this discrepancy between metadata and content is confusing another program (Adobe Forms) I want to feed my JPEGs to.  

To clarify, if I run:  
`./magick image.jpeg -set colorspace YCbCr new_image.jpeg`  
identify -verbose still reports the sRGB colorspace even though that command supposedly changes metadata.  

Requesting a conversion to YCbCr explicitly before the metadata change like so:  
`./magick image.jpeg -colorspace -set colorspace YCbCr new_image.jpeg`  
also gives the same result.

Interestingly, doing  
`./magick image.jpeg -colorspace CMYK image_new.jpeg`  
does show CMYK in the Colorspace field when identify'ing image_new.jpeg.  

So is there some way then to make the metadata correctly report YCbCr? Or is YCbCr assumed with sRGB? (which makes little sense to me as a novice)? If so, how can I verify that a JPEG reporting sRGB is indeed coded in YCbCr?

**Comment by snibgo** [comment, marked as answer]
Yes, JPEGs can also encode pixels as CMYK.

IM won't tell you how a _file_ is encoded. If you want to know that, use exiftool.

IM will tell you how an image stored in its _memory_ is encoded. When you use "-colorspace XXX", this changes how the pixels are encoded (by changing pixel values, and changing metadata). So, for example:
```
f:\web\im>%IMG7%magick toes.jpg -colorspace YCbCr info:
toes.jpg JPEG 267x233 267x233+0+0 8-bit YCbCr 22561B 0.000u 0:00.002

f:\web\im>%IMG7%magick toes.jpg -colorspace Lab info:
toes.jpg JPEG 267x233 267x233+0+0 8-bit CIELab 22561B 0.000u 0:00.001
```
"info:" reports the colorspace of the image that is in memory.

When IM writes a file, it knows that not all file formats can record all colorspaces. For example:
```
f:\web\im>%IMG7%magick toes.jpg -colorspace Lab x.jpeg

f:\web\im>exiftool x.jpeg
ExifTool Version Number         : 12.13
File Name                       : x.jpeg
Directory                       : .
File Size                       : 22 KiB
File Modification Date/Time     : 2022:03:18 12:38:04+00:00
File Access Date/Time           : 2022:03:18 12:38:04+00:00
File Creation Date/Time         : 2022:03:18 12:38:04+00:00
File Permissions                : rw-rw-rw-
File Type                       : JPEG
File Type Extension             : jpg
MIME Type                       : image/jpeg
JFIF Version                    : 1.01
Resolution Unit                 : cm
X Resolution                    : 118
Y Resolution                    : 118
Image Width                     : 267
Image Height                    : 233
Encoding Process                : Baseline DCT, Huffman coding
Bits Per Sample                 : 8
Color Components                : 3
Y Cb Cr Sub Sampling            : YCbCr4:4:4 (1 1)
Image Size                      : 267x233
Megapixels                      : 0.062
```
I have read the file toes.jpg, and converted that image to Lab, and saved the image to x.jpeg. Exiftool then tells me that x.jpeg is encoded as YCbCr. This has happened because IM knows that jpeg can't record Lab, so IM has converted the data to YCbCr.

> If so, how can I verify that a JPEG reporting sRGB is indeed coded in YCbCr?

To find how a file is encoded, use exiftool, not IM.
